README.md
# growthcurver
[![Travis-CI Build Status](https://travis-ci.org/sprouffske/growthcurver.svg?branch=master)](https://travis-ci.org/sprouffske/growthcurver)

Growthcurver is an R package that fits growth curve data to a standard form of the logistic equation common in ecology and evolution whose parameters (the growth rate, the initial population size, and the carrying capacity) provide meaningful population-level information with straight-forward biological interpretation. 

You can install the latest released version from CRAN from within R with
  ```R
  install.packages("growthcurver")
  ````

You can install the latest development version from github with
  ```R
  # install devtools first if you don't already have the package
  install.packages("devtools")

  # then install growthcurver
  devtools::install_github("sprouffske/growthcurver")
  ```

The paper describing Growthcurver is available [here](https://bmcbioinformatics.biomedcentral.com/articles/10.1186/s12859-016-1016-7).

## Using growthcurver

The easiest way to get started with growthcurver is to work through the examples in the vignette. In the vignette, you can find information on

* What your input data should look like
* How to use growthcurver to get summary metrics on a single growth curve sample
* How to use growthcurver to get summary metrics on an entire plate of growth curves
* What those metrics mean and some best practices for quality control

You can find the vignette [here](https://CRAN.R-project.org/package=growthcurver/vignettes/Growthcurver-vignette.html).

## A simple working example

This code loads the `growthcurver` package and some sample data. Then, it calls `SummarizeGrowth` to do the analysis. 
```R
library(growthcurver)                    # load the package
d <- growthdata                          # load some sample, simulated data
gc_fit <- SummarizeGrowth(d$time, d$A1)  # do the analysis
plot(gc_fit)                             # plot your data and the best fit
gc_fit                                   # view some returned metrics
```




NEWS.md
# growthcurver 0.3.1
This is a minor release to update Growthcurver for `R 4.0`.

# growthcurver 0.3.0

This is a minor release containing a bug fix and removing a dependency.

* Fixed bug in `SummarizeGrowthByPlate` that did not properly recognize the `time` and `blank` columns in some instances
* Updated `EmpiricalAreaUnderCurve` to compute the area under the curve using the trapezoid method, thereby removing the dependency on `trapz` from the orphaned `caTools` package 

# growthcurver 0.2.1

This is a minor release containing several bug fixes.

* Handled errors raised by `FitLogistic` when the model cannot be fit
* Added a `note` to the `gc_val` object and the output of 
  `SummarizeGrowthByPlate` to indicate failed or poor fits
* Prevented warnings raised by `TAtInflection` when $k < N_0$
* Updated the vignette to explain the new `note` value

# growthcurver 0.2.0

This is a minor release containing one new function to summarize the growth 
curves for many wells with a single command. Background correction has also
been added.

* Added `SummarizeGrowthByPlate` 
* Extended the `plot` method for `gc_fit` objects to take standard plot options
* Added background correction (option `bg_correct`)
* Updated the vignette to include the new options
* Set the default `bg_correct` option to `min`

# growthcurver 0.1.0

* Initial release


tests/testthat.R
library(testthat)
library(growthcurver)

test_check("growthcurver")


tests/testthat/test_NAtT.R
library(growthcurver)
context("Find n at time t (NAtT)")

test_that("NAtT returns the correct n at time t", {
  k_in <- 0.5
  n0_in <- 1e-5
  r_in <- 1.2
 
  expect_equal(n0_in,  
               growthcurver:::NAtT(k = k_in, n0 = n0_in, r = r_in, t = 0))
  expect_equal(k_in,  
               growthcurver:::NAtT(k = k_in, n0 = n0_in, r = r_in, t = 1000))
  expect_equal(round(0.2475303, 5),  
               round(growthcurver:::NAtT(k = k_in, n0 = n0_in, 
                                         r = r_in, t = 9), 5))
})



tests/testthat/test_DtAtT.R
context("Find doubling time at time t (DtAtT)")

test_that("DtAtT returns the correct doubling time at time t", {
  k_in <- 0.5
  n0_in <- 1e-5
  r_in <- 1.2
 
  expect_equal(round(0.577631, 5),  
               round(growthcurver:::DtAtT(k = k_in, 
                                          n0 = n0_in, 
                                          r = r_in, 
                                          t = 0), 5))
})



R/utils.R
#' Number of Cells at Time t
#'
#' This function gives the number of cells or absorbance (N) at time t when
#' the parameters to the logistic equation are K, N0, and r.
#' @param k       The carrying capacity
#' @param n0      The initial population size (absorbance or individuals)
#' @param r       The exponential "growth rate"
#' @param t       The time at which you want to know N
#' @return        The number of cells, or N, at time t
#' @export
NAtT <- function(k, n0, r, t) {
  return( k / (1 + ((k - n0) / n0) * exp(-r * t)))
}


# Slope At Time t
#
# This function gives the growth rate (or slope of the curve) when
# the parameters to the logistic equation are K, N0, and r.
# @param k       The carrying capacity
# @param n0      The initial population size (absorbance or individuals)
# @param r       The exponential "growth rate"
# @param t       The time at which you want to know the growth rate
# @return        The number of cells, or N, at time t
SlopeAtT <- function(k, n0, r, t) {
  n <- NAtT(k, n0, r, t)
  return(r * n * (k - n) / k)
}

# Fastest Doubling Time
#
# This function gives you the maximum doubling time (DT) assuming exponential
# growth.
# @param r       The exponential "growth rate"
# @return        The maximum doubling time
MaxDt <- function(r) {
  return(log(2) / r)
}


# Doubling (Generation) Time at Time t
#
# This function gives you the doubling time (DT) at time t when the parameters
# of the logistic equation are K, N0, and r.
# @param k       A single integer specifying the carrying capacity
# @param n0      The initial population size
#                (in either absorbance or individuals)
# @param r       The exponential "growth rate"
# @param t       The time at which you want to know the doubling time
# @return        The doubling time at time t
DtAtT <- function(k, n0, r, t) {
  n_t <- NAtT(k, n0, r, t)
  n_halft <- 0.5 * n_t
  return(( 1 / r) * log((n_t * (k - n_halft)) / ((k - n_t) * n_halft)))
}


# Time at Inflection Point
#
# This function returns the time of the inflection point
# of the logistic equation with parameters K, N0, and r.
# @param k       A single integer specifying the carrying capacity
# @param n0      The initial population size
#                (in either absorbance or individuals)
# @param r       The exponential "growth rate"
# @return        The time of the inflection point, which occurrs when the
#                the population size N reaches half its maximum value, K
TAtInflection <- function(k, n0, r) {
  if (n0 == 0) {
    warning("Initial population size (n0) cannot be 0.")
    return(0)
  }
  t_inflection <- log(abs(k - n0) / n0) / r
  return(t_inflection)
}


# Area Under the Logistic Curve
#
# This function gives you the area under the curve from time t_min to t_max,
# when the parameters of the logistic equation are K, N0, and r. This value
# essentially combines the lag phase, growth rate, and carrying capacity
# into a single value.
# @param k       The carrying capacity
# @param n0      The initial population size
#                (in either absorbance or number of individuals)
# @param r       The exponential "growth rate"
# @param t_min   The time from which you want to know the area under the curve
#                (e.g., from time = t_min to t_max)
# @param t_max   The time to which you want to know the area under the curve
#                (e.g., from time = t_min to t_max)
# @return        The area under the curve for logistic equation with the
#                given parameters, for the specificed time range
AreaUnderCurve <- function(k, n0, r, t_min = 0, t_max) {
  auc_l <- stats::integrate(function(x) NAtT(k, n0, r, x), t_min, t_max)
  return(auc_l)
}



# Area Under the Empirical Curve
#
# This function returns the empirical "area under the curve". It uses the input
# data to do so (rather than using the logistic fit).
# @param data_t    A vector of timepoints (data_n must also
#                  be provided and be the same length).
# @param data_n    A vector of cell counts or absorbance readings.
# @param t_trim    Add up the area under the curve from the beginning to
#                  t_trim. Defaults to 0, which means don't trim.
# @return          The area under the curve
EmpiricalAreaUnderCurve <- function(data_t, data_n, t_trim = 0) {
  # make sure that both inputs are vectors
  if (!is.vector(data_t) | !is.vector(data_n)) {
    stop("Error: The input data (data_t and data_n) must be vectors.")
  }
  if (!is.numeric(data_t) | !is.numeric(data_n)) {
    stop("Error: The input data (data_t and data_n) must be numeric.")
  }
  if (t_trim > 0) {
    idx_to_keep <- data_t <= t_trim                # keep the early measurements
  }
  else {
    idx_to_keep < rep(TRUE, length(data_t))       # keep all measurements
  }

  x <- data_t[idx_to_keep]
  y <- data_n[idx_to_keep]
  n <- length(x)

  auc_e <- sum((x[2:n] - x[1:n-1]) * (y[2:n] + y[1:n-1]) /  2)
  return(auc_e)
}


R/helpers.R
# Prints a gcvals object
# @param x         An object of class "gcvals"
# @param ...       Additional parameters
#' @export
print.gcvals <- function(x, ...) {
  cat(names(x[1:6]), sep = "\t")
  cat("\n")
  cat(c(mapply(x[1:2][], FUN = function(x) {round(x, 3)}),
        format(x[3], scientific = TRUE, digits = 1),
        mapply(x[4:5][], FUN = function(x) {round(x, 3)}),
        format(x[6], scientific = TRUE, digits = 1)),
      sep = "\t")
  cat("\n\n")

  cat(names(x[7:12]), sep = "\t")
  cat("\n")
  cat(c(mapply(x[7:8][], FUN = function(x) {round(x, 3)}),
        format(x[9], scientific = TRUE, digits = 1),
        mapply(x[10:12][], FUN = function(x) {round(x, 3)})),
      sep = "\t")
  cat("\n\n")

  cat(names(x[c(13:15)]), sep = "\t")
  cat("\n")
  cat(mapply(x[c(13:15)][], FUN = function(x) {round(x, 3)}), sep = "\t")
  cat("\n\n")
}

# Summarizes objects of class gcfit
# @param object         An object of class gcfit
# @param ...            Additional parameters
#' @export
summary.gcfit <- function(object, ...) {
  cat("Fit data to K / (1 + ((K - N0) / N0) * exp(-r * t))", "\n\n")
  print(object)
}


# Prints objects of class gcfit
# @param x         An object of class gcfit
# @param ...       Additional parameters
#' @export
print.gcfit <- function(x, ...) {

  # Show the fit
  cat("Fit data to K / (1 + ((K - N0) / N0) * exp(-r * t)):", "\n")
  cat("  ")
  cat("","K","N0", "r", sep= "\t")
  cat("\n")
  cat("  ")
  cat(c("val:", mapply(c(x$vals$k, x$vals$n0, x$vals$r),
                       FUN= function(x) {round(x, 3)})),
      sep="\t")
  cat("\n")
  cat("  ")
  cat("Residual standard error:", x$vals$sigma, "on", x$vals$df, "degrees of freedom")
  cat("\n\n")

  # Show other interesting metrics
  cat("Other useful metrics:\n")
  cat("  ")
  cat("DT", "1 / DT", "auc_l", "auc_e", sep = "\t")
  cat("\n")
  cat("  ")
  if (x$vals$note == "") {
    cat(c(round(x$vals$t_gen, 2),
          format(1 / x$vals$t_gen, scientific = TRUE, digits = 2),
          round(x$vals$auc_l, 2),
          round(x$vals$auc_e, 2)), sep = "\t")
  }
  else {
    cat("0", "0", "0", "0", sep = "\t")
    cat("\n\n")
    cat(c("Note: ", x$vals$note), sep = "")
  }
  cat("\n")
}

# Plots the data and the best fitting curve
# @param x         An object of class gcfit
# @param ...       Additional parameters
#' @export
plot.gcfit <- function(x, ...) {
  dots <- list(...)

  if (is.null(dots$xlab)) {
    dots$xlab <- "Time t"
  }
  if (is.null(dots$ylab)) {
    dots$ylab <- "Number N"
  }
  if (is.null(dots$pch)) {
    dots$pch <- 19
  }

  args_plot <- dots[setdiff(names(dots), "col")]
  args_plot$x <- x$data$t
  args_plot$y <- x$data$N
  args_plot$type <- "b"
  args_plot$ylim <- c(0, max(x$data$N))
    if (!is.null(dots$col)) {
    args_plot$col <- "black"
  }

  args_lines <- dots
  args_lines$x <- max(x$data$t) * (1 : 30) / 30
  args_lines$y <- NAtT(x$vals$k,
                       x$vals$n0,
                       x$vals$r,
                       args_lines$x)
  if (is.null(dots$col)) {
    args_lines$col <- "red"
  }

  old_par <- graphics::par(mfrow = c(1,1), mar = c(4,4,1,1))

  do.call(graphics::plot, args_plot)
  do.call(graphics::lines, args_lines)
  graphics::par(old_par)
}

#' Creates an object of type gcvals.
#'
#' Constructor function for the "gcvals" class. This object is most often
#' obtained when calling  \code{\link{SummarizeGrowth}} (it is the first
#' parameter in the \code{\link{gcvals}} object).
#' @param k         The carrying capacity parameter
#' @param k_se      The standard error of the carrying capacity parameter
#' @param k_p       The p value of the carrying capacity parameter
#' @param n0        The initial population size
#' @param n0_se     The standard error of the initial population size
#' @param n0_p      The p value of the initial population size
#' @param r         The growth rate
#' @param r_se      The standard error of the growth rate
#' @param r_p       The p value of the growthrate
#' @param sigma     Residual standard error from non-linear least squares fit
#'                  of the model to the data
#' @param df        Degrees of freedom
#' @param t_mid     The time at the inflection point of the logistic curve
#'                  (occurs at half of the carrying capacity)
#' @param dt        The maximum doubling time, obtained by evaluating the
#'                  the unrestrained growth of the population with growth rate r
#' @param auc_l     The area under the curve of the fitted logistic equation
#'                  from time 0 to time t
#' @param auc_e     The area under the curve of the measurements.
#' @param note      Feedback on common problems with fitting the logistic curve
#'                  to the data
#' @return          An object of class gcvals.
#' @export
gcvals <- function(k, k_se, k_p, n0, n0_se, n0_p, r, r_se, r_p, sigma, df,
                   t_mid, dt, auc_l, auc_e, note) {
  val.names <- c("k", "k_se", "k_p", "n0", "n0_se", "n0_p",
                 "r", "r_se", "r_p", "sigma", "df",
                 "t_mid", "dt", "auc_l", "auc_e", "note")
  vals <- stats::setNames(as.list(k, k_se, k_p, n0, n0_se, n0_p, r, r_se, r_p,
                           sigma, df, t_mid, dt,  auc_l, auc_e, note),
                   val.names)
}

#' Creates an object of class gcfit.
#'
#' This is a constructor function for the "gcfit" class. This class is most
#' often obtained as the return value when calling
#' \code{\link{SummarizeGrowth}}.
#' @param gc_vals   An object of class gcvals that contains the summarized
#'                  metrics from fitting the growth model to a set of
#'                  experimental observations. This is where the fitness proxy
#'                  parameters can be found. See \code{\link{gcvals}} for
#'                  more information the information found in this object.
#' @param log_mod    An object of class nlsModel that contains the results
#'                  of fitting the logistic growth model to the data
#' @param data_t    A numeric vector of times
#' @param data_n    A numeric vector of cell count or absorbance readings
#' @return          An object of class gcfit, which is a list of three objects,
#'                  that combines the parameters (vals = gc_vals,
#'                  model = log_mod, data = list(data_t, data_n))
#' @export
gcfit <- function(gc_vals, log_mod, data_t, data_n) {

  ret <- list("vals" = gc_vals, "model" = log_mod,
              "data"=list("t" = data_t, "N" = data_n))
  class(ret) <-"gcfit"
  return(ret)
}

#' Simulated growth curve data
#'
#' A dataset containing absorbance measurements over time of microbes
#' growing in a plate reader for 1 day. The growth curves for a whole plate are
#' included.
#'
#' @format A data frame with 145 observations and 97 variables:
#' \describe{
#'   \item{time}{time, in hours}
#'   \item{A1}{absorbance readings of well A1}
#'   \item{A2}{absorbance readings of well A2}
#'   \item{A3}{absorbance readings of well A3}
#'   \item{A4}{absorbance readings of well A4}
#'   \item{A5}{absorbance readings of well A5}
#'   \item{A6}{absorbance readings of well A6}
#'   \item{A7}{absorbance readings of well A7}
#'   \item{A8}{absorbance readings of well A8}
#'   \item{A9}{absorbance readings of well A9}
#'   \item{A10}{absorbance readings of well A10}
#'   \item{A11}{absorbance readings of well A11}
#'   \item{A12}{absorbance readings of well A12}
#'   \item{B1}{absorbance readings of well B1}
#'   \item{B2}{absorbance readings of well B2}
#'   \item{B3}{absorbance readings of well B3}
#'   \item{B4}{absorbance readings of well B4}
#'   \item{B5}{absorbance readings of well B5}
#'   \item{B6}{absorbance readings of well B6}
#'   \item{B7}{absorbance readings of well B7}
#'   \item{B8}{absorbance readings of well B8}
#'   \item{B9}{absorbance readings of well B9}
#'   \item{B10}{absorbance readings of well B10}
#'   \item{B11}{absorbance readings of well B11}
#'   \item{B12}{absorbance readings of well B12}
#'   \item{C1}{absorbance readings of well C1}
#'   \item{C2}{absorbance readings of well C2}
#'   \item{C3}{absorbance readings of well C3}
#'   \item{C4}{absorbance readings of well C4}
#'   \item{C5}{absorbance readings of well C5}
#'   \item{C6}{absorbance readings of well C6}
#'   \item{C7}{absorbance readings of well C7}
#'   \item{C8}{absorbance readings of well C8}
#'   \item{C9}{absorbance readings of well C9}
#'   \item{C10}{absorbance readings of well C10}
#'   \item{C11}{absorbance readings of well C11}
#'   \item{C12}{absorbance readings of well C12}
#'   \item{D1}{absorbance readings of well D1}
#'   \item{D2}{absorbance readings of well D2}
#'   \item{D3}{absorbance readings of well D3}
#'   \item{D4}{absorbance readings of well D4}
#'   \item{D5}{absorbance readings of well D5}
#'   \item{D6}{absorbance readings of well D6}
#'   \item{D7}{absorbance readings of well D7}
#'   \item{D8}{absorbance readings of well D8}
#'   \item{D9}{absorbance readings of well D9}
#'   \item{D10}{absorbance readings of well D10}
#'   \item{D11}{absorbance readings of well D11}
#'   \item{D12}{absorbance readings of well D12}
#'   \item{E1}{absorbance readings of well E1}
#'   \item{E2}{absorbance readings of well E2}
#'   \item{E3}{absorbance readings of well E3}
#'   \item{E4}{absorbance readings of well E4}
#'   \item{E5}{absorbance readings of well E5}
#'   \item{E6}{absorbance readings of well E6}
#'   \item{E7}{absorbance readings of well E7}
#'   \item{E8}{absorbance readings of well E8}
#'   \item{E9}{absorbance readings of well E9}
#'   \item{E10}{absorbance readings of well E10}
#'   \item{E11}{absorbance readings of well E11}
#'   \item{E12}{absorbance readings of well E12}
#'   \item{F1}{absorbance readings of well F1}
#'   \item{F2}{absorbance readings of well F2}
#'   \item{F3}{absorbance readings of well F3}
#'   \item{F4}{absorbance readings of well F4}
#'   \item{F5}{absorbance readings of well F5}
#'   \item{F6}{absorbance readings of well F6}
#'   \item{F7}{absorbance readings of well F7}
#'   \item{F8}{absorbance readings of well F8}
#'   \item{F9}{absorbance readings of well F9}
#'   \item{F10}{absorbance readings of well F10}
#'   \item{F11}{absorbance readings of well F11}
#'   \item{F12}{absorbance readings of well F12}
#'   \item{G1}{absorbance readings of well G1}
#'   \item{G2}{absorbance readings of well G2}
#'   \item{G3}{absorbance readings of well G3}
#'   \item{G4}{absorbance readings of well G4}
#'   \item{G5}{absorbance readings of well G5}
#'   \item{G6}{absorbance readings of well G6}
#'   \item{G7}{absorbance readings of well G7}
#'   \item{G8}{absorbance readings of well G8}
#'   \item{G9}{absorbance readings of well G9}
#'   \item{G10}{absorbance readings of well G10}
#'   \item{G11}{absorbance readings of well G11}
#'   \item{G12}{absorbance readings of well G12}
#'   \item{H1}{absorbance readings of well H1}
#'   \item{H2}{absorbance readings of well H2}
#'   \item{H3}{absorbance readings of well H3}
#'   \item{H4}{absorbance readings of well H4}
#'   \item{H5}{absorbance readings of well H5}
#'   \item{H6}{absorbance readings of well H6}
#'   \item{H7}{absorbance readings of well H7}
#'   \item{H8}{absorbance readings of well H8}
#'   \item{H9}{absorbance readings of well H9}
#'   \item{H10}{absorbance readings of well H10}
#'   \item{H11}{absorbance readings of well H11}
#'   \item{H12}{absorbance readings of well H12}
#' }
"growthdata"


R/fit-logistic.R
# Fits a logistic curve to data.
#
# This function fits a logistic curve to the supplied data, namely
# n(t) = K / (1 + ( (K - N0) / N0) * exp(-r * t), where
# N(t) is the number of cells (or density) at time t,
# K is the carrying capacity,
# N0 is the initial cell count or density, and
# r is the "growth rate".
# @param data_t    A vector of timepoints (data_n must also
#                  be provided and be the same length).
# @param data_n    A vector of cell counts or absorbance readings.
# @return          An object of class nls.
# @keywords        growth curves
FitLogistic <- function(data_t, data_n) {

  # make sure that the inputs are valid
  if (!is.vector(data_t) | !is.vector(data_n)) {
    stop("Error: The input data (data_t and data_n) must be vectors.")
  }
  if (!is.numeric(data_t) |!is.numeric(data_n)) {
    stop("Error: The input data (data_t and data_n) must be numeric.")
  }
  if (length(data_t) != length(data_n)) {
    stop("Error: The input data (data_t and data_n) must have the same length.")
  }

  # put together data
  d <- data.frame(cbind(data_t, data_n))
  names(d) <- c("t", "n")

  # make some guesses for the initial parameter values
  k_init <- max(data_n)   # carrying capacity is near the max
  n0_init <- min(data_n[data_n > 0])  # init population size is near the min

  # make an initial estimate for r
  glm_mod <- stats::glm(n / k_init ~ t,
                        family = stats::quasibinomial("logit"),
                        data = d)

  r_init <- stats::coef(glm_mod)[[2]]   # slope
  if (r_init <= 0) {
    # the slope should only be positive for a growing culture, so default
    # to something small
    r_init <- 0.001
  }

  suppressWarnings(
    nls_mod <- tryCatch(
                 minpack.lm::nlsLM(n ~ k / (1 + ( (k - n0) / n0) * exp(-r * t)),
                                   start = list(k = k_init,
                                                n0 = n0_init,
                                                r = r_init),
                                   control = list(maxiter = 500),
                                   lower = c(stats::median(data_n), 0, 0),
                                   upper = c(Inf, max(data_n), Inf),
                                   data = d),
                 error = function(e) {
                   stop("Error: Growthcurver FitLogistic cannot fit data.")
                 }
               )
  )
  return(nls_mod)
}


R/summarize-growth.R
#' Summarize Growth Curves
#'
#' This function finds the parameters that describe the input data's growth.
#' It does so by fitting the logistic curve to your growth curve measurements.
#'
#' The logistic curve equation is
#' \deqn{N_t = \frac{N_0 K} {N_0 + (K-N_0)e^{-rt}}}{N_t = N_0 K / (N_0 + (K - N_0) exp(-rt))}
#' where \eqn{N_t} is the number
#' of cells (or the absorbance reading) at time t, \eqn{N_0} is the initial
#' cell count (or absorbance reading), K is the carrying capacity, and r is the
#' growth rate.
#'
#' The fitness proxies returned are the parameters of the logistic equation
#' and the area under the curve (a measure that integrates the effects
#' of \eqn{N_0}, K, and r). See \code{\link{gcfit}} for more documentation on these.
#' @param data_t     A vector of timepoints (data_n must also
#'                   be provided and be the same length).
#' @param data_n     A vector of cell counts or absorbance readings.
#' @param t_trim     Measurements taken after this time should not be included
#'                   in fitting the curve. If stationary phase is variable,
#'                   this may give you a better fit. A value of 0 means no
#'                   trimming. Defaults to no trimming (0).
#' @param bg_correct The background correction method to use. No background
#'                   correction is performed for the default "none". Specifying
#'                   "min" subtracts the smallest value in a column from all the
#'                   rows in that column, and specifying "blank" subtracts
#'                   the values from the blank vector from the data_n vector.
#' @param blank      A vector of absorbance readings from a blank well
#'                   (typically contains only media) used for background
#'                   correction. The corresponding blank value is subtracted
#'                   from the data_n vector for each timepoint. Defaults to NA.
#' @return           An object of type gcfit containing the "fitness" proxies,
#'                   as well as the input data and the fitted model.
#' @seealso
#' See the accompanying Vignette for an example of how to use and interpret
#' SummarizeGrowth. \url{https://CRAN.R-project.org/package=growthcurver/vignettes/Growthcurver-vignette.html}.
#'
#' See also \code{\link{gcfit}}.
#' @examples
#' # We can check that the parameters that are found are the same
#' # as we use to generate fake experimental data. To do so, let's first
#' # generate the "experimental" data using the logistic equation,
#' # e.g., absorbance readings from a single well in a plate reader over time.
#'
#' k_in <- 0.5   # the initial carrying capacity
#' n0_in <- 1e-5 # the initial absorbance reading
#' r_in <- 1.2   # the initial growth rate
#' N <- 50       # the number of "measurements" collected during the growth
#'               # curve experiment
#'
#' data_t <- 0:N * 24 / N   # the times the measurements were made (in hours)
#' data_n <- NAtT(k = k_in, n0 = n0_in, r = r_in, t = data_t) # the measurements
#'
#' # Now summarize the "experimental" growth data that we just generated
#' gc <- SummarizeGrowth(data_t, data_n)
#'
#' # Get the possible metrics for fitness proxies
#' gc$vals$r           # growth rate is a common choice for fitness
#' gc$vals$t_gen       # doubling time, or generation time, is also common
#' gc$vals$k
#' gc$vals$n0
#' gc$vals$auc_l
#' gc$vals$auc_e
#' gc$vals$t_mid
#'
#' # Compare the data with the fit visually by plotting it
#' plot(gc)
#'
#' @export
SummarizeGrowth <- function(data_t, data_n, t_trim = 0,
                            bg_correct = "min", blank = NA) {

  # make sure that both inputs are vectors
  if (is.list(data_t) == TRUE) {
    tryCatch( {data_t <- unlist(data_t)},
              error = function(e) {stop("data_t is not a vector")}
            )
  }
  if (is.list(data_n) == TRUE) {
    tryCatch( {data_n <- unlist(data_n)},
              error = function(e) {stop("data_n is not a vector")}
            )
  }
  stopifnot(is.vector(data_t))
  stopifnot(is.vector(data_n))

  # make sure that the inputs are valid
  if (!is.numeric(data_t) |!is.numeric(data_n)) {
    stop("Error: The input data (data_t and data_n) must be numeric.")
  }
  if (length(data_t) != length(data_n)) {
    stop("Error: The input data (data_t and data_n) must have the same number of rows")
  }

  # make sure that the background correction method is valid
  if (!bg_correct %in% c("none", "min", "blank")) {
    stop(paste0(bg_correct, "is not a valid option for bg_correct"))
  }

  # check for correctness of the blank (background correction) vector
  if (bg_correct == "blank") {
    if (is.list(blank) == TRUE) {
      tryCatch( {blank <- unlist(blank)},
                error = function(e) {stop("blank is not a vector")}
      )
      stopifnot(is.vector(blank))

      if (!is.numeric(blank)) {
        stop("Error: The blank data must be numeric.")
      }
      if (length(blank) != length(data_n)) {
        stop("Error: The input data (data_n) and the background correction data (blank) must have the same number of rows.")
      }
    }
  }

  # check t_trim parameter and set dependent variables
  if (t_trim > 0) {
    t_max <- t_trim
    data_n <- data_n[data_t < t_trim]
    data_t <- data_t[data_t < t_trim]
    if (bg_correct == "blank") {
      blank <- blank[data_t < t_trim]
    }
  } else {
    t_max <- max(data_t, na.rm=TRUE)
  }

  #do the background correction
  if (bg_correct == "blank") {
    data_n <- data_n - blank
    data_n[data_n < 0] <- 0    # ensure readings are at least 0
  }
  else if (bg_correct == "min") {
    data_n <- data_n - min(data_n)
  }

  tryCatch(
    # code block
    {log_mod = FitLogistic(data_t, data_n)},
    # error handling block
    error = function(e) {}
    )

  # the data did not fit a logistic model
  if (exists("log_mod") == FALSE) {
    vals <- c(0, 0, 0, 0, 0, 0,
              0, 0, 0, 0, 0,
              0, 0, 0, 0)
  }
  else {
    p <- summary(log_mod)$coefficients
    k <- p[1]
    k_se <- p[4]
    k_p <- p[10]
    n0 <- p[2]
    n0_se <- p[5]
    n0_p <- p[11]
    r <- p[3]
    r_se <- p[6]
    r_p <- p[12]

    # get the inflection point, DT, auc, sigma, df
    t_inflection <- TAtInflection(k, n0, r)

    DT <- MaxDt(r)
    sigma <- summary(log_mod)$sigma
    df <- summary(log_mod)$df[2]

    auc_l <- AreaUnderCurve(k, n0, r, 0, t_max)$value
    auc_e <- EmpiricalAreaUnderCurve(data_t, data_n, t_max)
    vals <- c(k, k_se, k_p, n0, n0_se, n0_p,
              r, r_se, r_p, sigma, df,
              t_inflection, DT, auc_l, auc_e)
  }

  val_names <- c("k", "k_se", "k_p", "n0", "n0_se", "n0_p",
                 "r", "r_se", "r_p", "sigma", "df",
                 "t_mid", "t_gen", "auc_l", "auc_e")
  vals <- stats::setNames(as.list(vals), val_names)
  class(vals) <- "gcvals"

  if (exists("log_mod") == FALSE) {
    vals$note <- "cannot fit data"
    log_mod <- ""
  }
  else if (k < n0) {
    vals$note <- "questionable fit (k < n0)"
  }
  else if (t_inflection < 0) {
    vals$note <- "questionable fit"
  }
  else {
    vals$note <- ""
  }

  ret <- list("vals" = vals, "model" = log_mod,
              "data"=list("t" = data_t, "N" = data_n))
  class(ret) <- "gcfit"
  return(ret)
}


